Fig 8 and AWS update

ngreenwald opened the pull request:
Refactors/updates the Figure8 and aws pipelines to require less human intervention and work with Caliban data. 

Also rearranged and consolidated the core functions into more understandable pieces now that this repo has a clearly defined purpose. 

I think I am now fairly happy with the basic functionality of all of the code here. Some logical next steps are increasing test coverage (see #74) and adding in error checking throughout the pipeline (see #78).

@willgraf, I'd be particularly interested in your input on the current test suite, as well as how to go about testing the upload/download functions, although that may fit better into a subsequent PR.

Closes #72, closes #75, closes #77.

willgraf commented on caliban_toolbox/aws_functions.py:
```diff
@@ -65,7 +67,110 @@ def connect_aws():
+        s3.download_file(Bucket='caliban-output', Key=img_path, Filename=save_path)
+
+
+# old version of function: can delete once we're sure we won't use pipeline anymore
+def aws_upload_deprecated(bucket_name, aws_folder, folder_to_upload, include_context):
```
When will we know we can remove this?

ngreenwald replied:
@geneva-miller what do you think?

geneva-miller replied:
With zoom and the nicer interface for browser Caliban, I don't think we'll need to go back to the fig8 segmentation tool. We don't have in-progress fig8 tool jobs to support so I think we can go ahead and deprecate it now.

geneva-miller replied:
I think with zoom and the layout updates to browser Caliban we don't have much reason to use the fig8 segmentation tool anymore. We don't have any jobs of that kind in progress so we may as well deprecate it now.

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
+    saved_files = os.listdir(crop_dir)
+
+    # loop through all npz files
+    for fov_idx, fov_name in enumerate(fov_names):
+        crop_idx = 0
+        for row in range(num_row_crops):
+            for col in range(num_col_crops):
+                for slice in range(num_slices):
```
There must be a better way to do this. I am not wild about 4 nested for loops.

ngreenwald replied:
The data structure that this is looping through is an array with (at an abstract level) the following format: [FOVs, x/y crops, z slices]. For each FOV, we have images from distinct x/y crops. For each x/y crop, we can have images from distinct z slices. 

We currently loop over the 'row' crops and 'column' crops, which is sort of a meaningless distinction. Those two loops could easily be collapsed into a singe loop over all crops. 

However, I'm not sure how to get below 3 nested loops. We really do have 3 different levels of data.

geneva-miller replied:
I agree with Noah on this one. Big datasets get chopped up across each dimension to make them reasonable to annotate, and we end up with many smaller npzs to put back together. There's not a way for us to get around having fovs * rows * cols * slices number of npzs to load, and I think the very nested structure makes the most sense here, even if in general we try to avoid nested loops.

ngreenwald replied:
I modified it to now have only 3 for loops. This actually cleaned up some of the other code as well.

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
+        raise ValueError('Stacks must be combined before stitching can occur')
+
+    # loop through all crops in the stack for each image
+    for fov in range(fov_len):
+        for stack in range(stack_len):
+            crop_counter = 0
+            for row in range(len(row_starts)):
+                for col in range(len(col_starts)):
```
Ew! again, this looks very slow.

willgraf replied:
Any hope for this loop?

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
+
+    _, stack_len, _, _, row_len, col_len, _ = log_data['original_shape']
+    stitched_xr = xr.DataArray(data=stitched_images,
+                               coords=[fov_names, range(stack_len), range(1), range(1), range(row_len), range(col_len),
+                                       ['segmentation_label']], dims=['fovs', 'stacks', 'crops', 'slices',
+                                                                      'rows', 'cols', 'channels'])
```
this is pretty complicated. maybe you can break out the logic a bit so someone could understand what's happening?

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
+    stitched_slices[:, slice_start_indices[last_idx]:slice_end_indices[last_idx], :, 0, ...] = slice_stack[:, :slice_len, :, last_idx, ...]
+
+    stitched_xr = xr.DataArray(stitched_slices,
+                               coords=[fov_names, range(stack_len), range(crop_num), range(1), range(row_len),
+                                       range(col_len), ['segmentation_label']],
+                               dims=['fovs', 'stacks', 'crops', 'slices', 'rows', 'cols', 'channels'])
```
Here too I think a couple extra lines could make this a bit easier to understand.

ngreenwald replied:
Agreed, new changes make this much cleaner! Good catch

willgraf commented on .travis.yml:
```diff
@@ -17,7 +17,7 @@ install:
 
 script:
-  - python -m pytest caliban_toolbox tests
+  - python -m pytest caliban_toolbox
```
Why not include a pep8 and coverage test here too?

`python -m pytest --pep8 --cov=caliban_toolbox caliban_toolbox`

willgraf commented on caliban_toolbox/aws_functions.py:
```diff
@@ -0,0 +1,166 @@
+    subfolders = '__'.join(subfolders)
+
+    # optional flags to add to URL
+    optional_flags = np.any([pixel_only, label_only, rgb_mode])
+    if optional_flags:
+        optional_url = "?"
+        if pixel_only:
+            optional_url += "&pixel_only=true"
+        if label_only:
+            optional_url += "&label_only=true"
+        if rgb_mode:
+            optional_url += "&rgb=true"
```
This clearly works, but I think what you really want to do here is [urlencode your request arguments](https://stackoverflow.com/questions/5607551/how-to-urlencode-a-querystring-in-python). There's a bit of trickyness in getting this to import in 2 and 3, which I think we [resolved in the redis-consumer](https://github.com/vanvalenlab/kiosk-redis-consumer/blob/b3d6c17e8b0f37eb609a39e89f768949fb0900ed/redis_consumer/consumers/base_consumer.py#L680).

willgraf commented on caliban_toolbox/aws_functions.py:
```diff
@@ -0,0 +1,166 @@
+                       ExtraArgs={'ACL': 'public-read', 'Metadata': {'source_path': img_path}})
+        print('\n')
+
+        url = "https://caliban.deepcell.org/{0}__{1}__{2}__{3}__{4}".format('caliban-input',
+                                                                            'caliban-output',
+                                                                            subfolders, stage, img)
+
+        if optional_flags:
+            url += optional_url
```
I would prefer to always use the same quote char (so single quotes for the URL).

If you opt to try the urlencoding, you can just update `url` to be:

```python
url = 'https://caliban.deepcell.org/{0}__{1}__{2}__{3}__{4}?{5}'.format(
    'caliban-input',
    'caliban-output',
    subfolders, stage, img,
    url_encoded_dict
)
```

willgraf commented on caliban_toolbox/figure_eight_functions.py:
```diff
@@ -0,0 +1,193 @@
+        key: API key to access Figure 8 account
+    """
+
+    url = "https://api.figure-eight.com/v1/jobs/{job_id}/upload.json?key={api_key}&force=true"
+    url = url.replace('{job_id}', str(job_id))
+    url = url.replace('{api_key}', key)
```
Oh another opportunity for urlencoding!

willgraf commented on caliban_toolbox/log_file.py:
```diff
@@ -0,0 +1,244 @@
+
+        optional_flags = np.any(pixel_only, label_only, rgb_mode)
+
+        if optional_flags:
+            optional_url = "?"
+            if pixel_only:
+                optional_url += "&pixel_only=true"
+            if label_only:
+                optional_url += "&label_only=true"
+            if rgb_mode:
+                optional_url += "&rgb=true"
+
+        new_filepath = "https://caliban.deepcell.org/{0}__{1}__{2}__{3}__{4}".format(input_bucket,
+                                                                                     output_bucket,
+                                                                                     subfolders,
+                                                                                     next_stage,
+                                                                                     filename)
```
And like above, another opportunity for urlencoding.

willgraf commented on caliban_toolbox/relabel_test.py:
```diff
@@ -89,7 +88,8 @@ def test_relabel_all_frames():
 #     input_data = np.load('tests/caliban_toolbox/stack_j046_i003_all_channels.npz')['annotated']
-#     true_data = np.load('tests/caliban_toolbox/stack_j046_i003_all_channels_relabeled.npz')['annotated']
+#     true_data =
+#         np.load('tests/caliban_toolbox/stack_j046_i003_all_channels_relabeled.npz')['annotated']
```
You can break up anything as long as its in parenthesis:

```python
true_data = np.load('tests/caliban_toolbox/'
                    'stack_j046_i003_all_channels_relabeled.npz')['annotated']
```

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -0,0 +1,703 @@
+    if blank_labels == 'separate':
+        os.makedirs(os.path.join(save_dir, 'separate'))
+
+    # loop through all crops in all images
+    for fov in range(fov_len):
+        for crop in range(num_crops):
+            for slice in range(num_slices):
```
Glad we got it down to 3 from 4. Instead of using a triple for loop, I suggest looking into [`itertools.combinations`](https://docs.python.org/3/library/itertools.html#itertools.combinations). It is both easier to read and [apparently faster](https://stackoverflow.com/questions/37354372/itertools-vs-nested-loops-performance).

ngreenwald replied:
It's so beautiful

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -0,0 +1,703 @@
+                      num_slices, row_crop_size, col_crop_size, 1))
+    saved_files = os.listdir(crop_dir)
+
+    # loop through all npz files
+    for fov_idx, fov_name in enumerate(fov_names):
+        for crop in range(num_crops):
+            for slice in range(num_slices):
```
See above comment about `itertools.combinations`.

geneva-miller commented on caliban_toolbox/reshape_data.py:
```diff
@@ -0,0 +1,703 @@
+    Returns:
+        string: formatted file name
+
+    Raises:
```
This doesn't appear to raise an error if a match isn't found, only if too many matches are found

ngreenwald commented on caliban_toolbox/log_file.py:
```diff
@@ -0,0 +1,244 @@
+        os.chmod(log_dir, mode)
+
+    # save csv file
+    dataframe.to_csv(os.path.join(log_dir, 'upload_log.csv'), index=False)
```
Totally agree about overwriting. Currently it will overwrite. Next PR when I add in the multiple_stages script I'm going to enforce sequential CSV names and check that file doesn't exist. I just haven't worked out the exact syntax yet.

ngreenwald replied:
I wasn't sure what to call this. Something more informative would be great that signifies this file contains the necessary information for creating a figure8 job.

ngreenwald commented on caliban_toolbox/log_file.py:
```diff
@@ -0,0 +1,244 @@
+
+    # pull identifier info from csv_data, this will be used in filename saving
+    # note: not suited for job reports that have a mix of identifiers
+    identifier = csv_data['identifier'].values[0]
```
Yeah, that whole function needs to be updated; left it as is for now

ngreenwald commented on caliban_toolbox/log_file.py:
```diff
@@ -0,0 +1,244 @@
+    identifier = csv_data['identifier'].values[0]
+
+    next_job_df = pd.DataFrame(data=data, index=range(len(filepath_list)))
+    next_csv_name = os.path.join(csv_dir, '{0}_{1}_upload.csv'.format(identifier, next_stage))
```
Yeah, I like this! That would take care of the issue of overwriting the csv if it exists already

willgraf commented on caliban_toolbox/figure_eight_functions.py:
```diff
@@ -0,0 +1,194 @@
+        key: API key to access Figure 8 account
+    """
+
+    url = "https://api.appen.com/v1/jobs/{job_id}/upload.json?key={api_key}&force=true"
+    url = url.replace('{job_id}', str(job_id))
+    url = url.replace('{api_key}', key)
```
`.format()` seems like a better option than the calling `replace()`.

ngreenwald replied:
Agreed. I'm gonna look into url formatting for all of it. Seeing now if capitalization of true/false matters for caliban

ngreenwald commented:
Alright, I think I addressed all of your comments. I'm still up in the air about what to call the csv file that gets created as part of each figure8 job, but I'm definitely going to incorporate Geneva's suggestion about numbering them according to stage

willgraf commented on caliban_toolbox/aws_functions.py:
```diff
@@ -0,0 +1,156 @@
+import threading
+import re
+
+from urllib.parse import urlencode
```
I expect this to have some issues with py2 vs py3 compatibility. You can check the redis-consumer to see how we handled it before.

ngreenwald replied:
So currently the toolbox is only being built against 3.6. 

Do you think we should expand? I understand the rationale for deepcell-tf, which is much more widely used. Not so sure it makes sense here, especially since they just released the last ever 2.X build.

willgraf replied:
Good point. I forgot which project I was in.  Probably no need to build for py2, but we should probably build it against a few python3 versions (3.5, 3.6, 3.7?)

willgraf commented on caliban_toolbox/log_file_test.py:
```diff
@@ -0,0 +1,24 @@
+import tempfile
```
Add the license/copyright comment here as well.

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -0,0 +1,706 @@
+        os.makedirs(os.path.join(save_dir, 'separate'))
+
+    # for each fov, loop through 2D crops and 3D slices
+    for indices in product(range(fov_len), range(num_crops), range(num_slices)):
+        fov, crop, slice = indices
```
can be even simpler:

```python
for fov, crop, slice in product(...):
```

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -0,0 +1,706 @@
+    saved_files = os.listdir(crop_dir)
+
+    # for each fov, loop over each 2D crop and 3D slice
+    for indices in product(range(fov_len), range(num_crops), range(num_slices)):
+        fov, crop, slice = indices
```
can be even simpler:

```python
for fov, crop, slice in product(...):
```

willgraf commented on caliban_toolbox/reshape_data.py:
```diff
@@ -0,0 +1,706 @@
+    print('annotated shape: {}'.format(annotated_data.shape))
+    print('num rows, num cols {} {}'.format(len(row_starts), len(col_starts)))
+    # for each fov and stack, loop through rows and columns of crop positions
+    for indices in product(range(fov_len), range(stack_len),
+                           range(len(row_starts)), range(len(col_starts))):
+
+        fov, stack, row, col = indices
```
can be even simpler:

```python
for fov, stack, row, col in product(...):
```

willgraf commented on .travis.yml:
```diff
@@ -6,8 +6,8 @@ git:
 
-python:
-  - 3.6
+python:  
```
Looks like extra trailing whitespace could be removed.

ngreenwald replied:
Good catch, thanks!

willgraf commented on .travis.yml:
```diff
@@ -6,8 +6,8 @@ git:
 
-python:
-  - 3.6
+python: 
```
```suggestion
python:
```

willgraf replied:
Still some missing whitespace.

ngreenwald replied:
Got it, I'm gonna wait to see what we decide re: url_encoding before merging this in, in case we decide to modify that functionality

willgraf commented:
I learned a bit about how Flask casts the request args. `request.args` returns a `MultiDict` which has a [`.get` method](https://github.com/pallets/werkzeug/blob/6accf1f85f4dd5662e4c7f8ab6b115ed13037374/src/werkzeug/datastructures.py#L235).  This allows a `type` parameter, which we've set to boolean.

```python
# what we do in the route
label_only = request.args.get('label_only', type=bool)

# which is the same as
label_only = reqargs.get('label_only')  # string by default, as its from a URL
label_only = bool(label_only)
```

Which is a problem because:

```python
assert bool('True') is True and bool('False') is True
```

---

It looks like `flask_restful` uses `flask_restful.inputs.boolean` to [fix this problem](https://github.com/flask-restful/flask-restful/issues/397).

```

ngreenwald commented:
Is there a reason this is only a problem for label_only and rgb_mode or pixel_only?

willgraf commented:
> Is there a reason this is only a problem for label_only and not rgb_mode or pixel_only?

I am very surprised that it does evaluate to False if you say `rgb=false` in the request. Can you confirm that? It may be that we have some logic in the background that is evaluating it differently.

ngreenwald commented:
Nope, you're right. I only ever tried it with =True.
It has the same issue

ngreenwald commented:
Okay great, in that case I'll leave the logic here as is and we you guys can update flask?
